### ISSUE TITLE
Simplify internals to prep for 1.0 release

### ISSUE THREAD

**Issue by jbogaardt** [casact/chainladder-python#601]
On the path towards 1.0 there are several refactors worth considering that would simplify the code base and bring it more into software development best practices.

# Backend support
To migrate to an arrow backend, we should minimize existing backend support to just numpy/sparse. 

cupy and dask were always elective and not well maintained. They add unnecessary complexity to the project and should be removed.

# Clarify Triangle public API
The Estimators occasionally manipulate the private internals of the Triangle class. This makes the Estimator code heavily coupled to the internals of the Triangle making refactors of Triangle internals difficult. We need to force etimators to only use the public API.  This may include adding additional support properties and methods to the Triangle class that the estimators can call on as opposed to direct manipulation of the underlying triangle structure.

None of these should change any of the exiting public API of the package and all test should continue to pass.

**Comment by henrydingliu** [comment]
yo, I think 2D is the solution to all our problems. I'm so sick of all the grain issues. i'm one more grain issue away from rewriting the triangle internals to force storing at OMDM, and only regraining for estimation and display. In 2D, we can keep claim-level accident date and dates of transactions. then only map them to o and d grain at run time and aggregate for estimation or display. basically 2D enables sparse (if I'm understanding the purpose of sparse correctly) support by default.

**Comment by henrydingliu** [comment]
Alas we don't have to choose between actuarial techniques or data management. Here is an excerpt from Clark 2003, which we reference in the package but is also on the exam 7 syllabus currently. According to him, Arrow is integral to more techniques. 

<img width="447" height="292" alt="Image" src="https://github.com/user-attachments/assets/d5855610-f211-4d9e-97b1-24115f539bed" />

Let me think about how we can do some performance testing next.

**Comment by genedan** [comment]
I need close out a bunch of browser windows full of ideas so I'm recording them here:

Have any of you explored [JAX](https://docs.jax.dev/en/latest/notebooks/thinking_in_jax.html)? It looks like an all-in-one library with CPU, GPU, parallel processing, and sparse support, with a numpy-compatible API, which I thought could cut down on trying to maintain 4 separate backends. The main downside I see is that their sparse support is still listed as experimental. I think we can keep an eye on this one to see where it goes.

Since Polars is a dataframe and not a multidimensional array library, I suppose it's off the table as a backend option. I think there might be some performance improvements in select areas, such as I/O like when data are ingested into a DataFrame during the `Triangle` `__init__`, or when they are output as DataFrames.

**Comment by henrydingliu** [comment]
Arrow is also columnar. Migrating to Arrow/Polars essentially involves melting 4D triangles all into 2D, i.e. origin and development as explicit columns, such as most of our sample csv files. 

My position on this has been evolving. I no longer think migrating the entire backend would make much sense. Or rather, I think it's the current assumption of inseparability between triangle-development_transformer-chainladder_predictor that needs to evolve. for example, without making any change to the current backend, we can easily allow `chainladder().predict` to accept a dataframe with an origin column and a development column (i.e. calculate age based on the origin and development, then join the cdf_ based on age). and once we do this, my sense is that we no longer need sparse, since the motivation for sparse is to keep claim-level ultimate, rather than deriving claim-specific ldf_. and if we can accept that estimation of ldf_ and application of ldf_ doesn't to be based on triangles exclusively, then we can really get freaky. 

basically, our superclass doesn't need to generalize 4D arrays. it just needs to generalize any data analysis library that has groupby and join, as groupby and join are the only two operations we need to perform triangle display, and triangle estimation. we no longer need to declare a triangle to do anything. we can relegate `triangle` to be a display format and an accelerator in numpy. loss data is stored in their native format. they are displayed in a triangle format at runtime using the groupby functionality provided by their native library; ldf_ is estimated directly on the underlying claims dataframe by using join to match between successive development periods (or estimated in a 4D numpy array, which should be faster than any columnar estimation); cdf_ is calculated using group by on ldf_; cdf_ is applied using join again. other conceptual shifts with this approach:
- we essentially get rid of the notion of using the chainladder package to do something like `tri1 + tri2`. manipulate loss data in pandas or arrow or spark or whatever the user fancies
- we now only have a `cl.displayTriangle(data, origin_column_name, development_column_name)` method that takes a dataframe of any ilk and show it like a triangle. instead of `tri3 = tri1 + tri2`, it's `cl.displayTriangle(df1 + df2)`
- `cl.Development().fit(dataframe)` creates a triangle at runtime, either goes through the current weighted regression implementation OR go through a new implementation using joins, and lastly stores ldf_ as dataframe where age is the index. 

if you think about it, data scientists didn't require people to declare a tree or a network to train and infer using gbm or NN. that level of abstraction is hidden behind the implementation. triangle is a mental aid for the actuary to think about development pattern. it never needed to be more than that. and since all P&C claims systems store data in a columnar fashion, that what our estimators should train and infer on. 

to tie this back to the realities of this package, i guess i'm saying that i don't see the value of any backend work before 1.0, as the very notion that `Triangle` should be the central construct of this package seems outdated. by 2.0, i would hope that we are putting `Triangle` on deprecation warning, and mainly support estimators that directly work on columnar data.

**Comment by genedan** [comment]
I'm wondering how to actually handle the removal of the 'dims. They are arrays but the Triangle public dimensions aren't. Also in the case of `odims` the datetimes are also more precise. Since `index`, `columns`, `origin`, and `development` are dynamic properties of the Triangle, I think it would be more accurate to say that those are aliases for the 'dims rather than the other way around.

As I type this up, I'm having second thoughts about all those deprecation sub-issues I just created. Should we just convert the 'dims to be private attributes? Since they are all of type array, they do provide a consistent type to work with.

**Comment by genedan** [comment]
I'm digging around the Pandas code to see what it does. Both index and columns get initialized to a general axis class called [`AxisProperty`](
https://github.com/pandas-dev/pandas/blob/721ab260c67dd480e83d1510fa5ac08553536038/pandas/_libs/properties.pyx#L47-L69) which gets stored in something called a [`BlockManager`](https://github.com/pandas-dev/pandas/blob/721ab260c67dd480e83d1510fa5ac08553536038/pandas/core/internals/managers.py#L1097) which handles them. Anyway, that's where my brain started hurting and where I stopped looking, but it's pretty cool we're getting this far in the decision-making process for writing a package.

Having a generalized `TriangleAxis` class would make sense. It would have its own getter and setter so the axes would have a consistent API. But before moving in this direction we should confirm that it's actually needed, i.e., something needed beyond just setting the 'dims to private and using common ndarrays.

**Comment by henrydingliu** [comment]
>I think it would be more accurate to say that those are aliases for the 'dims rather than the other way around

as long as the alias is 1-to-1, then it wouldn't matter if the alias is the 'source of truth', we can make the alias into the 'source of truth' and the 'source of truth' into the alias. 

>Since they are all of type array, they do provide a consistent type to work with.

>Both index and columns get initialized to a general axis class called [AxisProperty](https://github.com/pandas-dev/pandas/blob/721ab260c67dd480e83d1510fa5ac08553536038/pandas/_libs/properties.pyx#L47-L69)

does it work for chainladder if we just had 4 of those axis classes?

**Comment by genedan** [comment]
> as long as the alias is 1-to-1, then it wouldn't matter if the alias is the 'source of truth', we can make the alias into the 'source of truth' and the 'source of truth' into the alias.

Which representation did you have in mind? I think they should be stored as Indexes. `columns` should be an `Index`, since the array could be extracted via `Traingle.columns.values`, and that would make the `vdims` array redundant, and we wouldn't have to store it as a property.

> does it work for chainladder if we just had 4 of those axis classes?

Pandas just has 2, so we might have more of a use for abstraction. But I still don't have a compelling reason to make `TriangleAxis` other than that it would be fun to try to do it. We could get started on phasing out the `dims` first, and if in doing that work we find a better reason to create a general axis class, we could, but if we don't, we could just hold off.

**Comment by genedan** [comment]
I did find a reason for why Pandas might have a general `AxisProperty` class. There are several Pandas indexes, `Index`, `MultiIndex`, `DateIndex`, `RangeIndex`, `PeriodIndex`, etc. The DataFrame index and columns can be any of these, so the generalization to `AxisProperty` allows that to happen.

Chainladder is stricter, with `Index` being a `DataFrame`, `columns` being an `Index`, origin being either a `Series` (pattern case) or a `PeriodIndex`, and `valuation` being a `DatetimeIndex`. Hypothetically, we could allow flexible index types on all the dimensions, but I think that would be a massive undertaking and a discussion for another day.

There might be some justification for `TriangleAxis` because of the two return types of `origin`, but an alternative I've been thinking about is that having a distinct `Pattern` class would make more sense, due to the numerous `if self.is_pattern:` statements throughout the library.

I also think the `Triangle.index` should really return an `Index` or `MultiIndex`, and not `DataFrame`, which would be more consistent with Pandas behavior:

```python
import pandas as pd
df = pd.DataFrame(
    data={'a': ['foo', 'bar'], 'b': ['baz', 'boo'], 'c': [1, 2]}
)
df = df.set_index(['a', 'b'])
print(type(df.index))
```

```
<class 'pandas.core.indexes.multi.MultiIndex'>
```

```python
import chainladder as cl

clrd = cl.load_sample('clrd')
print(type(clrd.index))
```

```
<class 'pandas.core.frame.DataFrame'>
```

**Comment by henrydingliu** [comment]
> Which representation did you have in mind? I think they should be stored as Indexes.

yes

> origin being either a Series (pattern case) or a PeriodIndex, and valuation being a DatetimeIndex

valuation is an implied dimension. development is the explicit dimenion and is a series :/

> but an alternative I've been thinking about is that having a distinct Pattern class would make more sense, due to the numerous if self.is_pattern: statements throughout the library

I've thought about that before. But it also feels a little contrived since all we use pattern to do is to multiply into triangles

> I also think the Triangle.index should really return an Index or MultiIndex, and not DataFrame, which would be more consistent with Pandas behavior:

I can't find the passage right now but I recall reading somewhere in the docs that John chose DataFrame over MultiIndex for some reason.
